### ISSUE TITLE
Wrong boxes location in train and val labels 

### ISSUE THREAD

**Issue by aguirrejuan** [ultralytics/ultralytics#18186]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train, Val

### Bug

Hi!

I am training an MRZ detector using the YOLOv11 OBB task. I am labeling the data with Label Studio and exporting it in the YOLOv8 OBB format. However, when I visualize the data, I notice some distortion in the bounding boxes. Some boxes appear to occupy the entire image.

<img width="688" alt="Screenshot 2024-12-11 at 11 51 46" src="https://github.com/user-attachments/assets/7c613488-1503-4d4d-887d-551e7443a369" />

### Environment

Ultralytics 8.3.49 🚀 Python-3.12.7 torch-2.5.1 CPU (Apple M3 Pro)

### Minimal Reproducible Example

```
model.train(cfg="config.yaml",
                      epochs=1, imgsz=1080)
```

```
task: obb
mode: train
model: yolo11n-obb.pt
data: mrz-detection/artifacts/data.yaml
epochs: 100
time: null
patience: 100
batch: 16
imgsz: 640
save: true
save_period: -1
cache: false
device: null 
workers: 8
project: null
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: true
cos_lr: false
close_mosaic: 4
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0 
val: true
split: val
save_json: false
save_hybrid: false
conf: null
iou: 0.7
max_det: 1
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: true
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: null
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
nbs: 64
hsv_h: 0.3
hsv_s: 0.7
hsv_v: 0.4
degrees: 180.0
translate: 0.3
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.0
bgr: 0.5
mosaic: 0.4
mixup: 0.0
copy_paste: 0.1
copy_paste_mode: flip
auto_augment: randaugment
erasing: 0.0
crop_fraction: 0.0
cfg: null
tracker: botsort.yaml
save_dir: runs/obb/train5

```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @aguirrejuan, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to submit your issue.

Based on your description, it seems you might be encountering a 🐛 bug related to bounding box distortions. To help us investigate further, we kindly ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) that clearly demonstrates the issue. This could include:
- A small, shareable subset of your labeled dataset that reproduces the bounding box problem.
- A colab or script example of how you are loading the data and visualizing the labels to reproduce the distortion.

We also recommend the following steps to troubleshoot in the meantime:
## Upgrade
Ensure your environment is using the latest `ultralytics` package and dependencies. This often resolves unexpected behaviors:
```bash
pip install - up τελευτα40 confirm other **)run.solve by switching visual to���] level

**Comment by Y-T-G** [comment]
That happens only on the plots. The actual training is not affected.

**Comment by aguirrejuan** [comment]
Hi! Thanks for your answer. Could you please explain how I can prove that this is the case?

**Comment by Y-T-G** [comment]
You can check the plotting function code.
https://github.com/ultralytics/ultralytics/blob/0504d3a8e004abf068d5c7b7bb3670602d6aa1ef/ultralytics/utils/plotting.py#L371

During training, the images are not tiled like in the plot. So there isn't a way for the box to bleed into another image.

**Comment by aguirrejuan** [comment]
The problem is not the bleeding of boxes into other images. The issues are related to the disappearance of the boxes and distortion as well. There are also boxes that significantly mismatch the objects.

**Comment by Y-T-G** [comment]
Can you try the latest version?

`pip install git+https://github.com/ultralytics/ultralytics`

**Comment by aguirrejuan** [comment]
I have already installed the latest version and tried again, but the problem persists.

`Ultralytics 8.3.51 🚀 Python-3.12.7 torch-2.5.1 CPU (Apple M3 Pro)`

**Comment by aguirrejuan** [comment]
I have installed the latest commit. It looks much better! Thanks!

**Comment by aguirrejuan** [comment]
Hi, even though there are improvements, there are still problems. In some images, the boxes disappear.

<img width="814" alt="Screenshot 2024-12-19 at 18 49 55" src="https://github.com/user-attachments/assets/eeb62258-321f-446a-a50e-1a3c3bdd8594" />

**Comment by Y-T-G** [comment]
@Laughing-q The issue with missing OBB boxes was also reported in https://github.com/ultralytics/ultralytics/issues/18288 recently

**Comment by Laughing-q** [comment]
> The subset is in the next file. However, I have blurred the images and removed some to create a smaller dataset.
> 
> [train.zip](https://github.com/user-attachments/files/18099443/train.zip)
> 
> ```
> path_dir: mrz-detection/artifacts
> train: ./train
> val: ./train
> 
> nc: 5
> 
> names:
>   0: MRVA
>   1: MRVB
>   2: TD1
>   3: TD2
>   4: TD3
> ```

@aguirrejuan Hi! this attached zip file is empty, can you upload a minimal reproducible image and label here so we are able to debug? Thanks

**Comment by chelseaztq** [comment]
Hello, I recently encountered a similar issue, but my problem is not related to the disappearance of labels. I believe the label disappearance is caused by the following line of code in the `RandomPerspective` class:

```python
# Make the bboxes have the same scale with new_bboxes
i = self.box_candidates(
    box1=instances.bboxes.T, box2=new_instances.bboxes.T, area_thr=0.01 if len(segments) else 0.10
)
```

My issue occurs after the labels are transformed and the following code is applied:

```python
# Update bboxes if there are segments.
if len(segments):
    bboxes, segments = self.apply_segments(segments, M)
```

And after clipping:

```python
# Clip
new_instances.clip(*self.size)
```

The edges of the bounding boxes, especially those with a large aspect ratio, are not clipped correctly.

As shown in the image below, the top is the original image, and the bottom is the cropped image. Both have the same coordinate shape of Nx100x2.
![image](https://github.com/user-attachments/assets/b9824bb9-f9c2-49bd-a95a-631d996e88be)

![image](https://github.com/user-attachments/assets/bfdadd6c-6a4c-42c9-b99f-033a0e94237c)

So when the following function is called later, it computes the bounding box around the segments and calculates the rotation angle:

```python
def xyxyxyxy2xywhr(x):
    """
    Convert batched Oriented Bounding Boxes (OBB) from [xy1, xy2, xy3, xy4] to [xywh, rotation]. Rotation values are
    returned in radians from 0 to pi/2.

    Args:
        x (numpy.ndarray | torch.Tensor): Input box corners [xy1, xy2, xy3, xy4] of shape (n, 8).

    Returns:
        (numpy.ndarray | torch.Tensor): Converted data in [cx, cy, w, h, rotation] format of shape (n, 5).
    """
    is_torch = isinstance(x, torch.Tensor)
    points = x.cpu().numpy() if is_torch else x
    points = points.reshape(len(x), -1, 2)
    rboxes = []
    for pts in points:
        # NOTE: Use cv2.minAreaRect to get accurate xywhr,
        # especially some objects are cut off by augmentations in dataloader.
        (cx, cy), (w, h), angle = cv2.minAreaRect(pts)
        rboxes.append([cx, cy, w, h, angle / 180 * np.pi])
    return torch.tensor(rboxes, device=x.device, dtype=x.dtype) if is_torch else np.asarray(rboxes)
```

If the current cropping method is used, the resulting bounding box may look like the one shown below. It appears to have a certain rotation. This is due to coordinate errors causing the bounding box to shift, not completely enclosing the target. I'm not sure if this will have a significant impact on the final training results.
![image](https://github.com/user-attachments/assets/576a883c-8522-4767-96c7-8455b7c18fe9)


So, I modified the code as follows:

```python
def xyxyxyxy2xywhr(x, w_img, h_img):
    # def xyxyxyxy2xywhr(x):
    """
    Convert batched Oriented Bounding Boxes (OBB) from [xy1, xy2, xy3, xy4] to [xywh, rotation]. Rotation values are
    returned in radians from 0 to pi/2.

    Args:
        x (numpy.ndarray | torch.Tensor): Input box corners [xy1, xy2, xy3, xy4] of shape (n, 8).

    Returns:
        (numpy.ndarray | torch.Tensor): Converted data in [cx, cy, w, h, rotation] format of shape (n, 5).
    """
    is_torch = isinstance(x, torch.Tensor)
    points = x.cpu().numpy() if is_torch else x
    points = points.reshape(len(x), -1, 2)
    rboxes = []
    for pts in points:
        # Filter out points near the edges (taking all four edges into account)
        pts = pts[(pts[:, 0] != w_img) & (pts[:, 1] != h_img) & (pts[:, 0] != 0) & (pts[:, 1] != 0)]  #TODO
        # NOTE: Use cv2.minAreaRect to get accurate xywhr,
        # especially when some objects are cut off by augmentations in dataloader.
        (cx, cy), (w, h), angle = cv2.minAreaRect(pts)
        rboxes.append([cx, cy, w, h, angle / 180 * np.pi])
    return torch.tensor(rboxes, device=x.device, dtype=x.dtype) if is_torch else np.asarray(rboxes)
```

When the coordinates are near the edges (considering all four edges), I discard these points and then generate the minimum enclosing rectangle.
![image](https://github.com/user-attachments/assets/f36612e0-0839-4ba4-81f2-c301bdbe0d03)


After discarding the edge coordinates, it seems that the angle of the minimum enclosing rectangle is calculated correctly, and the coordinates drawn on the image also maintain the same angle as before the cropping. This ensures that the bounding box encloses the target consistently with the pre-cropped image, rather than causing a rotation that results in extra background being included, as is happening now. This phenomenon is especially common with targets that have a very large aspect ratio. Specifically, the more the image is cropped, the larger the deviation of the ground truth box.

However, I'm not entirely sure if this improvement will be beneficial for training. In my current experiments, it does show an improvement, as the ground truth boxes are more aligned with my manual annotations.

**Comment by chelseaztq** [comment]
@aguirrejuan 
![image](https://github.com/user-attachments/assets/4e0a0cff-5fd7-4b19-b029-8380fa27744f)
I believe your labels are experiencing the same issue. It's clearly visible that the angle of this ground truth box is inconsistent with your actual annotated angle.

**Comment by Laughing-q** [comment]
@chelseaztq oh wow nice discovery! Could you submit a PR for this modification? Thanks

**Comment by aguirrejuan** [comment]
Hi! Regarding the issue with the disappearing images:

I noticed that the labels are still present on the images. If I’m correct, the labels are always placed in the bottom-right corner. In this case, it seems like the box shrinks and moves closer to the origin.

Interestingly, when I comment out the following lines:

https://github.com/ultralytics/ultralytics/blob/1d13575ba16623d711c682118ee118615383ba99/ultralytics/data/augment.py#L2064-L2066

all the missing boxes start reappearing.

**Comment by chelseaztq** [comment]
@aguirrejuan 

Commenting out that line of code is incorrect. Those two lines are meant to normalize the coordinates to a range between 0 and 1, preparing them for subsequent training.

As I mentioned before, you need to adjust:

# Make the bboxes have the same scale with new_bboxes
i = self.box_candidates(
    box1=instances.bboxes.T, box2=new_instances.bboxes.T, area_thr=0.01 if len(segments) else 0.10
)

There are three parameters for the thresholds, which should be clearly explained in the author’s comments:

	•	wh_thr
	•	ar_thr
	•	area_thr

After adjustment, you’ll notice that the ground truth boxes compressed into a single point or line are filtered out.

The following is GPT’s explanation.

Function Explanation: box_candidates

The box_candidates function evaluates whether bounding boxes remain valid after data augmentation, such as random perspective transformations. It ensures that only bounding boxes meeting specific criteria for width, height, aspect ratio, and area are preserved for further processing.

Purpose

	•	Why? During image augmentation, bounding boxes might become distorted, overly stretched, compressed into a line, or shrink to an invalid size.
	•	What does it do? This function filters out bounding boxes that no longer meet valid criteria after augmentation.

Parameters

	1.	box1 (numpy.ndarray)
	•	Original bounding boxes (before augmentation).
	•	Shape: (4, N) where N is the number of boxes.
	•	Format: [x1, y1, x2, y2] (coordinates of the top-left and bottom-right corners).
	2.	box2 (numpy.ndarray)
	•	Augmented bounding boxes (after transformation).
	•	Shape and format are the same as box1.
	3.	wh_thr (float)
	•	Minimum width and height threshold for bounding boxes.
	•	If a bounding box’s width or height falls below this threshold, it is considered invalid.
	•	Default: 2.
	4.	ar_thr (float)
	•	Maximum aspect ratio threshold.
	•	If the bounding box’s width-to-height ratio exceeds this value, it is considered invalid.
	•	Default: 100.
	5.	area_thr (float)
	•	Minimum area ratio threshold.
	•	Represents the ratio between the areas of the augmented bounding box (box2) and the original (box1).
	•	If the ratio falls below this threshold, the bounding box is considered invalid.
	•	Default: 0.1.
	6.	eps (float)
	•	A small epsilon value (1e-16) added to prevent division by zero errors during calculations.

Logic Breakdown

	1.	Calculate Width and Height for Both Sets of Boxes:

w1, h1 = box1[2] - box1[0], box1[3] - box1[1]
w2, h2 = box2[2] - box2[0], box2[3] - box2[1]

	•	w1, h1: Width and height of the original bounding boxes.
	•	w2, h2: Width and height of the augmented bounding boxes.

	2.	Calculate the Aspect Ratio (ar):

ar = np.maximum(w2 / (h2 + eps), h2 / (w2 + eps))

	•	Aspect ratio is calculated as the larger value between width/height and height/width.
	•	This ensures overly stretched (thin or elongated) boxes are flagged as invalid.

	3.	Filter Based on Criteria:

return (w2 > wh_thr) & (h2 > wh_thr) & (w2 * h2 / (w1 * h1 + eps) > area_thr) & (ar < ar_thr)

	•	(w2 > wh_thr) & (h2 > wh_thr): Ensures the bounding box’s width and height meet the minimum threshold.
	•	(w2 * h2 / (w1 * h1 + eps) > area_thr): Ensures the area of the augmented box is not excessively reduced compared to the original.
	•	(ar < ar_thr): Ensures the aspect ratio does not exceed the defined threshold.

	4.	Return a Boolean Array:
	•	Each element corresponds to one bounding box.
	•	True indicates the bounding box meets all criteria.
	•	False indicates it does not.

Example Usage

import numpy as np

# Original bounding boxes
box1 = np.array([[0, 0, 100, 100], [0, 0, 50, 50]]).T
# Augmented bounding boxes
box2 = np.array([[10, 10, 90, 90], [5, 5, 45, 45]]).T

# Call the method
candidates = box_candidates(box1, box2)
print(candidates)

Output:

[True, True]

	•	Both bounding boxes are valid after augmentation and pass the filtering criteria.

Key Takeaways:

	•	Ensures bounding boxes remain valid after augmentation.
	•	Prevents distorted or overly reduced bounding boxes from being used in training.
	•	Thresholds (wh_thr, ar_thr, area_thr) can be fine-tuned based on the dataset and augmentation methods.

This function plays a critical role in maintaining the quality of bounding box annotations throughout the data preprocessing pipeline.

**Comment by aguirrejuan** [comment]
Hi! @chelseaztq
Yes, you are correct—commenting out those lines is not a solution at all. **I just believe it might provide some insight to help locate the problem**.

Thank you for sharing the code for the possible solution!

**Comment by aguirrejuan** [comment]
Hi,

I have an update regarding the disappearance of the bounding boxes.

The issue appears to be related to the normalization of the xywhr boxes (from pixel coordinates to a normalized [0,1] range). [The cv2.minAreaRect function occasionally swaps the width (w) and height (h)](https://github.com/opencv/opencv/issues/19472). This can result in cases where either the width or height exceeds the expected tolerance.

For example, if a box has a large width and a very small height (almost the size of the entire image), the width may become significantly larger than the height. This produces a normalized width greater than one. When this happens, the normalization fails because it does not meet the required conditions in the if statement.

https://github.com/ultralytics/ultralytics/blob/51026a9a4a0b5e1ea8a1c76b16f8de12b9a71530/ultralytics/utils/plotting.py#L1100-L1102

Can this cause problems during training as well? @Y-T-G  @Laughing-q 

![](https://raw.githubusercontent.com/zytx121/image-host/main/imgs/opencv.png)

**Comment by Laughing-q** [comment]
@aguirrejuan ah I see I did encounter the same situation before and that's why I updated to `boxes[:, :4].max() <= 1.1`(originally it's 1.01) to mitigate the issue, and if that's the case, no need to worry about the training, this only affects the visualization.

**Comment by aguirrejuan** [comment]
Hi!

Understood. However, I have used segmentation models to accomplish the same task, and the performance is significantly better compared to using OBB. This might be due to the specific nature of my problem.

**Comment by glenn-jocher** [comment]
Thank you for sharing your findings! Segmentation models can indeed perform better for certain tasks where precise pixel-level annotations are crucial. If segmentation aligns better with your dataset and objectives, it’s a great choice. You can always explore integrating both approaches for complementary results.
